File: backend/services/network.js
import _ from "lodash";
import axios from "axios";

import { api } from "../utils/controller-api.js";
import { db } from "../utils/db.js";
import { defaultRulesSource } from "../utils/constants.js";

export async function getNetworkAdditionalData(data) {
  let additionalData = db
    .get("networks")
    .find({ id: data.id })
    .get("additionalConfig");

  if (!additionalData.value()) {
    createNetworkAdditionalData(data.id);
  }

  delete data.rulesSource;
  delete data.objtype;
  delete data.revision;
  delete data.remoteTraceLevel;
  delete data.remoteTraceTarget;

  return {
    id: data.id,
    clock: new Date().getTime(),
    ...additionalData.value(),
    config: data,
  };
}

export async function getNetworksData(nwids) {
  const prefix = "/controller/network/";
  const links = nwids.map((nwid) => prefix + nwid);

  const multipleRes = await axios
    .all(links.map((l) => api.get(l)))
    .then(
      axios.spread(function (...res) {
        return res;
      })
    )
    .catch(function () {
      return [];
    });

  let data = Promise.all(
    multipleRes.map((el) => {
      return getNetworkAdditionalData(el.data);
    })
  );

  return data;
}

export async function createNetworkAdditionalData(nwid) {
  const saveData = {
    id: nwid,
    additionalConfig: {
      description: "",
      rulesSource: defaultRulesSource,
      tagsByName: {},
      capabilitiesByName: {},
    },
    members: [],
  };

  db.get("networks").push(saveData).write();
}

export async function updateNetworkAdditionalData(nwid, data) {
  let additionalData = {};

  if (data.hasOwnProperty("description")) {
    additionalData.description = data.description;
  }
  if (data.hasOwnProperty("rulesSource")) {
    additionalData.rulesSource = data.rulesSource;
  }
  if (data.hasOwnProperty("tagsByName")) {
    additionalData.tagsByName = data.tagsByName;
  }
  if (data.hasOwnProperty("capabilitiesByName")) {
    additionalData.capabilitiesByName = data.capabilitiesByName;
  }

  if (additionalData) {
    db.get("networks")
      .filter({ id: nwid })
      .map("additionalConfig")
      .map((additionalConfig) => _.assign(additionalConfig, additionalData))
      .write();
  }
}

export async function deleteNetworkAdditionalData(nwid) {
  db.get("networks").remove({ id: nwid }).write();
}

File: backend-py/services/network.py
import sys, os, time
sys.path.append(os.getcwd())
from constants.constants import *

from vars import DARWIN, LINUX
sys_vars = DARWIN() if os.uname().sysname == "Darwin" else LINUX()


class NETWORK():
    """Classe um die Anfragen rund ums Networking zu handeln"""

    def getNetworksData(self, nwids : list) -> list:
        networks = []
        for nwid in nwids:
            controllerResponse = zerotier_api.make_request("GET", f"/controller/network/{nwid}").json()
            networks.append(self.getNetworkAdditionalData(controllerResponse))

        return networks

    def getNetworkAdditionalData(self, data : dict) -> dict:
        if db.read_networks_value(data.get("id")) == []:
            self.createNetworkAdditionalData(data.get("id"))

        additionalData = db.read_networks_value(data.get("id"))[0].get("additionalConfig")

        del data["rulesSource"]
        del data["objtype"]
        del data["revision"]
        del data["remoteTraceLevel"]
        del data["remoteTraceTarget"]

        networkData = {
            "id" : data.get("id"),
            "type" : "Network",
            "clock" : time.time(),
            "config" : data,
        }

        networkData = db.merge_two_dicts(networkData,additionalData)

        return networkData

    def createNetworkAdditionalData(self, nwid : str) -> None:
        saveData = {
            "id": nwid,
            "additionalConfig": {
            "description": "",
            "rulesSource": sys_vars.default_rule_source,
            "tagsByName": {},
            "capabilitiesByName": {},
                },
            "members": [],
            }
        
        db.write_value("networks", saveData)

    def updateNetworkAdditionalData(self, nwid : str, data : dict) -> None:
        additionalData = {}

        if (data.get("name")): 
            additionalData["name"] = data.get("name")

        if (data.get("description")): 
            additionalData["description"] = data.get("description")
        
        if (data.get("rulesSource")): 
            additionalData["rulesSource"] = data.get("rulesSource")
        
        if (data.get("tagsByName")): 
            additionalData["tagsByName"] = data.get("tagsByName")
        
        if (data.get("capabilitiesByName")): 
            additionalData["capabilitiesByName"] = data.get("capabilitiesByName")

        if (additionalData):
            db.update_network(nwid, additionalData)


        